doc/en/man/blockHashRead.1
.\" Do *not* edit this file; it was automatically generated by ej2man
.\" Look for a name.ej file with the same name as this filename
.\"
.\" Process this file with the following (replace filename.1)
.\" preconv < filename.1 | nroff -man -Tutf8
.\"
.\" Last updated 2022-12-18
.\"
.TH BLOCKHASHREAD 1 "Dec 2022" BLOCKHASHREAD "blockHashRead reference"
.\" We don't want hyphenation (it's too ugly)
.\" We also disable justification when using nroff
.\" Due to the way the -mandoc macro works, this needs to be placed
.\" after the .TH heading
.hy 0
.if n .na
.\"
.\" We need the following stuff so that we can have single quotes
.\" In both groff and other UNIX *roff processors
.if \n(.g .mso www.tmac
.ds aq \(aq
.if !\n(.g .if '\(aq'' .ds aq \'

  
.SH "NAME"
.PP
blockHashRead - Read a block hash file 
.SH "DESCRIPTION"
.PP
blockHashRead is a stand alone command line tool which converts a 
.I "block hash file"
in to a list of hostnames. This way, a binary block hash file can be 
converted in to an ASCII list of host names, edited, and then converted 
back in to a binary block hash file with the 
.B "blockHashMake"
utility. 
.PP
A block hash file uses a special binary format for storing a list of 
blocked host names. 
.SH "USAGE"
.PP
blockHashRead is invoked as follows:

.nf
blockHashRead --dump bigBlock.bin 
.fi

Replace "bigBlock.bin" with the filename for the block hash file. 
.PP
Doing this will output, on standard output, a list of host names in the 
block hash file. Each line will contain a single host name. When 
compiled for *NIX, the output will use *NIX line feeds; the Windows 
port of blockHashRead uses DOS line feeds. 
.PP
blockHashRead can be invoked with a single "--help" or "--version" 
command line argument (e.g. "blockHashRead --version") which will 
output the version number of blockHashRead and provide basic usage 
information. 
.SH "HOST LIST FORMAT"
.PP
After being invoked, blockHashRead writes a list of host names to the 
standard output. The format is a single host name per line of input, 
such as the following:

.nf
porn.example.com 
naughty.foo 
evil.host.invalid 
.fi

Each line is a host name. 
.PP
blockHashRead has no support for Punycode. Please use another program 
to convert international domain names with non-ASCII characters in to 
their non-punycode representation if seeing correct international 
domain names is desired. 
.SH "LIMITATIONS"
.PP
The block hash format that blockHashRead looks at is a 32-bit format, 
and the resulting block hash file should be under 2,147,483,648 bytes 
in size. This is a limitation of around 30 million host names. 
.SH "LEGAL DISCLAIMERS"
.PP
THIS SOFTWARE IS PROVIDED BY THE AUTHORS \(aq\(aqAS IS\(aq\(aq AND ANY 
EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE 
IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR 
PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHORS OR CONTRIBUTORS 
BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR 
CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF 
SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR 
BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, 
WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR 
OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF 
ADVISED OF THE POSSIBILITY OF SUCH DAMAGE. 
.PP
This is a project developed on a strictly volunteer, non-commercial 
basis. It has been developed outside the course of a commercial 
activity, developed entirely in the Americas (i.e. 
.IR "outside of Europe" ")"
and therefore is not subject to the restrictions or conditions of the 
proposed EU Cyber Resilience Act. Someone selling a product that uses 
any component of this may be subject to this act and may need to handle 
any and all necessary compliance. 
.SH "AUTHORS"
.PP
Sam Trenholme (https://www.samiam.org) is responsible for this program 
and man page.  



CHANGELOG
MaraDNS changelog

   maradns-3.5.0020:
   This is a stable release of MaraDNS:

     * New parameter: source_ip4, to specify the source IP when sending a
       query to an upstream or authoritative DNS server.
     * Makefile.centos8 file for Deadwood renamed Makefile.Ubuntu2004 (i.e.
       it’s now a Makefile for Ubuntu 20.04)

     (2021-05-??)

   maradns-3.5.0019:
   This is a stable release of MaraDNS:

     * One line change to zoneserver.c to make it work better with systemd
     * Synthetic IP generator example (e.g. 10.1.2.3.ip4.internal 
       resolves to 10.1.2.3) added to coLunacyDNS documentation

     (2021-03-16)

   maradns-3.5.0018:
   This is a stable release of MaraDNS:

     * coLunacyDNS updated to 1.0.010 (Ubuntu 20.04 testing found a
       minor select() bug)
     * Since RedHat has broken their pinky promise to support CentOS 8
       until 2029, MaraDNS has moved from CentOS 8 to Ubuntu 20.04 LTS
       (RPM .spec files removed; Docker container now runs Ubuntu 20.04)

     (2020-12-20)

   maradns-3.5.0017:
   This is a stable release of MaraDNS:

     * coLunacyDNS update: We can now specify the “Authoritative”
       and “Recursion available” flags in the reply.
     * coLunacyDNS update: We can now specify a TTL for the reply,
       to be anywhere from 0 seconds (do not cache) to a little over
       90 days.
     * coLunacyDNS version updated to be 1.0.009.

     (2020-10-02)

   maradns-3.5.0016:
   This is a stable release of MaraDNS:

     * Unstable mmLunacyDNS code removed from tree (coLunacyDNS can do
       anything mmLunacyDNS could do)
     * coLunacyDNS bug fixes: We return with an error if the Lua code
       attempts to return an invalid IPv4 address to the client (before,
       the code incorrectly returned 255.255.255.255)
     * coLunacyDNS returns helpful errors if the processQuery return 
       value is invalid in various ways.
     * coLunacyDNS now has 100%* testing coverage. *Some sanity tests
       which protect coLunacyDNS from security threats which can not
       be readily reproduced are disabled in testing mode.
     * coLunacyDNS is now at version 1.0.008

     (2020-09-01)

   maradns-3.5.0015:
   This is a stable release of MaraDNS:

     * coLunacyDNS is now a stable release (1.0.007) with a full
       SQA testing suite and well over 90% code coverage in its tests.
     * mmLunacyDNS has been removed; coLunacyDNS can do everything
       mmLunacyDNS could do, and this saves me the bother of
       maintaining two code bases.
     * askmara now compiles with IPv6 support (the code has been
       there, but is finally getting enabled)

     (2020-08-29)

   maradns-3.5.0014:
   This is a stable release of MaraDNS (note that coLunacyDNS and
   mmLunacyDNS are unstable):

     * coLunacyDNS’s Lua script can now specify IPv6 addresses in
       standard “colon” format, e.g. co1Data="2001:db8::1"
     * coLunacyDNS now handles ANY (and HINFO) queries as per RFC8482
     * coLunacyDNS documentation updates: Various cleanup.  Also,
       the example coLunacyDNS .lua files now return “not there”
       when we ask for a hostname which does not have a given record.

     (2020-08-20)

   maradns-3.5.0013:
   This is a stable release of MaraDNS (note that coLunacyDNS and
   mmLunacyDNS are unstable):

     * coLunacyDNS now has support for binding to IPv6 addresses.  Both
       the *NIX (Linux) and the Windows32 binary can bind to an IPv6
       socket.
     * Some other bug fixes and cleanup, mainly with coLunacyDNS.

     (2020-08-19)

   maradns-3.5.0012:
   This is a stable release of MaraDNS (note that coLunacyDNS and
   mmLunacyDNS are unstable): 

     * mmLunacyDNS security fix: We now use a secure hash compression
       function (HalfSipHash-1-3) for string hashing.
     * coLunacyDNS: hash compression function updated from 64-bit
       SipHash-2-4 to 32-bit HalfSipHash-1-3.  Compile time warnings
       removed from code.
     * lunacy: The code by default now uses HalfSipHash-1-3 for string
       hash compression.  Default compile optimization is now -O3
 
     (2020-08-12)

   maradns-3.5.0011:
   This is a stable release of MaraDNS:

     * min_ttl parameter added; this is the minimum time we keep a 
       record in the cache (in seconds)
     * Deadwood now compiles with IPv6 support by default.  For
       systems without IPv6 support, -DNOIP6 can be set when
       compiling Deadwood.
     * Automated tests now all run inside of Podman (Docker) container
       and all pass.  Tests are now completely automated, and can run
       from cron (and can be adapted to run inside Jenkins). 

     (2020-08-10)

   maradns-3.5.0010:
   This is a stable release of MaraDNS:

     * Hotfix: coLunacyDNS no longer fails after 20 calls to 
       processQuery() (we now properly clean the main stack before 
       calling processQuery() in a co-routine).
     * Security update: MaraDNS, Deadwood, and Duende now default to
       the user ID 707 instead of 99/66.  This minimizes the chances
       of the user used by MaraDNS being used by other processes, 
       which could be a security leak under some circumstances.  The
       problem with running multiple services as "nobody" is that
       the "nobody" account is only as secure as the least secure
       service running as that account.
     * coLunacyDNS feature update: coLunacyDNS can now open and
       read files (for security reasons, only in the same directory
       coLunacyDNS is running in).  In addition, the code to implement 
       IPv6 sockets is well under way.

     (2020-08-06)

   maradns-3.5.0009:
   This is a stable release of MaraDNS:

     * Add new program: coLunacyDNS.  This is a DNS server which runs a
       Lua function every time it gets a DNS query.  It uses Lua
       threads ("co-routines") to have a function which can get a
       DNS packet from an upstream server and return the result for

deadwood-github/tools/blockhash/blockHashRead.c
/* Copyright (c) 2022 Sam Trenholme
 *
 * TERMS
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * This software is provided 'as is' with no guarantees of correctness or
 * fitness for purpose.
 */

#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>
#include <fcntl.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#ifdef MINGW
#include <io.h>
#endif // MINGW


typedef struct {
  uint8_t *block;
  uint32_t max;
  uint32_t sipKey1;
  uint32_t sipKey2;
  int32_t hashSize;
} blockHash;

blockHash *globalBlockHash;
uint32_t globalSipKey1 = 0;
uint32_t globalSipKey2 = 0;

// The numbers are big endian
uint32_t read32bitNumber(uint8_t *block, uint32_t offset, uint32_t max) {
  uint32_t out = 0;
  if(offset + 3 >= max) {
    return 0;
  }
  out |= block[offset] << 24;
  out |= (block[offset + 1] & 0xff) << 16;
  out |= (block[offset + 2] & 0xff) << 8;
  out |= block[offset + 3];
  return out;
}
  
uint32_t HalfSip13(uint8_t *str, int32_t l, 
                   uint32_t sipKey1, uint32_t sipKey2) {
  if(str == NULL) { 
    return 0;
  }
  size_t offset = 0;
  uint32_t v0, v1, v2, v3, m;
  int shift = 0, round = 0;

  v0 = sipKey1;
  v1 = sipKey2;
  v2 = v0 ^ 0x6c796765;
  v3 = v1 ^ 0x74656462;
  m = 0;
  while(offset <= l) {
    if(offset < l) {
      m |= (uint32_t)(str[offset] & 0xff) << shift;
      shift += 8;
    }
    while(shift >= 32 || offset == l) { // "while" to avoid goto
      if(offset == l && shift != 32) {
        m |= (uint64_t)(l & 0xff) << 24;
        offset++;
      }
      shift = 0;
      v3 ^= m;

      v0 += v1;
      v1 = (v1 << 5) | (v1 >> 27);
      v1 ^= v0;
      v0 = (v0 << 16) | (v0 >> 16);
      v2 += v3;
      v3 = (v3 << 8) | (v3 >> 24);
      v3 ^= v2; v0 += v3;
      v3 = (v3 << 7) | (v3 >> 25);
      v3 ^= v0; v2 += v1;
      v1 = (v1 << 13) | (v1 >> 19);
      v1 ^= v2;
      v2 = (v2 << 16) | (v2 >> 16);

      v0 ^= m;
      shift = 0;
      m = 0;
    }
    offset++;
  }
  v2 ^= 255;
  for(round = 0; round < 3; round++) {
    v0 += v1;
    v1 = (v1 << 5) | (v1 >> 27);
    v1 ^= v0;
    v0 = (v0 << 16) | (v0 >> 16);
    v2 += v3;
    v3 = (v3 << 8) | (v3 >> 24);
    v3 ^= v2; v0 += v3;
    v3 = (v3 << 7) | (v3 >> 25);
    v3 ^= v0; v2 += v1;
    v1 = (v1 << 13) | (v1 >> 19);
    v1 ^= v2;
    v2 = (v2 << 16) | (v2 >> 16);
  }
  return v1 ^ v3;
}

/* Convert a string in to a DNS name.  For example,
 * " www.example.com\n" becomes "\3www\7example\3com\0".
 * The string is altered by this process.  The first character
 * in the string will be overwritten with the length of the first
 * DNS label; the input should be a string in the form <space>
 * followed by the DNS name in question, followed by a \n or space
 * This changes the string in place.
 */
int16_t dnsConvertString(int32_t len, uint8_t *str) {
  int dnsPoint = 0;
  int dnsLength = 0;
  int counter;
  if(len > 1040) {
    return 0; // Error
  }
  for(counter = 1; counter < len; counter++) {
    if(dnsPoint >= len || counter >= len) {
      return 0; // Error
    }
    if(str[counter] == '.' || str[counter] == '\n' ||
       str[counter] == ' ') {
      dnsLength = counter - dnsPoint - 1;
      if(dnsLength >= 0 && dnsLength < 64) {
        str[dnsPoint] = dnsLength;
      } else {
        return 0; // Error
      }
      if(dnsLength == 0) {
        return dnsPoint + 1; // A 0-length for a DNS label is end of string
      }
      dnsPoint = counter;
    }
  }
  dnsLength = counter - dnsPoint - 1;
  if(dnsLength >= 0 && dnsLength < 64) {
    str[dnsPoint] = dnsLength;
  }
  return len;
}

/* Given an ASCII string, convert it in to a DNS string.  The DNS string
 * is described in the dnsConvertString above.  This makes a *copy* of
 * the string in question.  olen is altered to have the length of the
 * output string */
uint8_t *ASCII2DNS(char *str, int32_t *olen) {
  uint8_t *out = NULL; 
  int32_t l;
  if(str == NULL || olen == NULL) { return NULL; }
  l = strlen(str);
  out = malloc(l + 3);
  if(out == NULL) { return NULL; }
  out[0] = ' ';
  out[l + 1] = ' ';
  if(strncpy(out + 1, str, l) == NULL) { free(out); return NULL; }
  *olen = dnsConvertString(l + 2, out);
  if(*olen == 0) { free(out); return NULL; }
  return out;
}
  
// See if a black has a given binary string (string str, length len)
// 1: Yes, it does
// 0: No, it does not
// -1: An error happened when trying to find the string
int blockHasString(blockHash *b, uint8_t *str, int32_t len) {
  uint32_t sipHashBucket;
  uint32_t offset;
  if(b == NULL || str == NULL) {
    return -1; // Error
  }
  sipHashBucket = HalfSip13(str, len, b->sipKey1, b->sipKey2) % b->hashSize;
  offset = read32bitNumber(b->block, 16 + (4 * sipHashBucket), b->max); 
  if(offset == 0) { 
    return 0; // Not found: No strings for this hash bucket
  }
  if(offset >= b->max - 1) { // - 1 because we will read 16-bit number
    return -1; // Error
  }
  while(offset < b->max) {
    int32_t stringLen, strOffset;
    uint32_t strEnd;
    stringLen = b->block[offset];
    if(stringLen == 0xff) { // End of hash chain
      return 0; // Not found: End of hash bucket chain reached
    }
    offset++;
    stringLen <<= 8;
    if(offset >= b->max - 1) { return -1; } // - 1 to avoid 2nd bounds check
    stringLen |= b->block[offset];
    if(stringLen < 0 || stringLen > 0xff00) { return -1; }
    offset++; 
    strEnd = offset + stringLen;
    if(strEnd > b->max) { return -1; }
    if(stringLen != len) { // Different length, try next string
      offset = strEnd;
      continue;
    }
    strOffset = 0;
    while(stringLen > 0) {
      if(b->block[offset] != str[strOffset]) { // Different string, try next
        offset = strEnd;
        break;
      }
      stringLen--;
      strOffset++;
      offset++;
    }
    if(stringLen == 0) {
      return 1; // Strings match, found
    }
  }
  return -1; // We should never get here, error   
}

// Read a file and make a blockHash structure
blockHash *makeBlockHash(char *filename) {
  blockHash *out = NULL;
  int fileDesc;
  out = malloc(sizeof(blockHash));
  if(out == NULL) { return NULL; }
  struct stat get;
  if(stat(filename,&get) == -1) {
    return NULL; 
  }
  out->max = get.st_size;
  if(out->max < 16) { free(out); return NULL; }
  out->block = malloc(out->max + 3);
  if(out->block == NULL) { free(out); return NULL; }
  fileDesc = open(filename, 0);
  if(fileDesc == -1) { free(out->block); free(out); return NULL; }
#ifdef MINGW
  setmode(fileDesc, O_BINARY);
#endif // MINGW
  if(read(fileDesc, out->block, out->max) != out->max) {
    free(out->block);
    free(out);
    return NULL;
  }
  if(out->block[0] != 0 || out->block[1] != 'D' || out->block[2] != 'w' 
     || out->block[3] != 'B') { free(out->block); free(out); return NULL; }
  out->sipKey1 = read32bitNumber(out->block,4,out->max);
  out->sipKey2 = read32bitNumber(out->block,8,out->max);
  out->hashSize = read32bitNumber(out->block,12,out->max);
  return out;
}

// Print out all of the strings in a given block list until the length
// is the 0xff terminator
int dumpStrings(uint8_t *block, uint32_t offset, uint32_t max) {
  int32_t len;
  while(offset <= max) {
    int32_t stringLen;
    uint32_t strEnd;
    int thisLabelLen; // DNS length label
    stringLen = block[offset];
    if(stringLen == 0xff) { // End of hash chain
      return 0; // End of hash bucket chain reached
    }
    offset++;
    stringLen <<= 8;
    if(offset >= max - 1) { return -1; } // - 1 to avoid 2nd bounds check
    stringLen |= block[offset];
    if(stringLen < 0 || stringLen > 0xff00) { return -1; }
    offset++; 
    strEnd = offset + stringLen;
    if(strEnd > max) { return -1; }
    thisLabelLen = -1;
    while(stringLen > 0) {
      if(thisLabelLen <= 0) {
        if(block[offset] == 0) { 
          offset = strEnd;
          break;
        }
        if(thisLabelLen == 0) {
          printf(".");
        }
        thisLabelLen = block[offset];
      } else {
        printf("%c",block[offset]);
        thisLabelLen--;
      }
      stringLen--;
      offset++;
    }
#ifndef MINGW
    printf("\n");
#else
    printf("\r\n");
#endif // MINGW
  } 
  return 0;
}

// Print out all of the DNS names in a blockHash
int dumpContents(char *filename) {
  int32_t counter;
  globalBlockHash = makeBlockHash(filename);
  if(globalBlockHash == NULL) {
    printf("Error reading %s\n",filename);
    return 1; // Error
  }
  for(counter = 0; counter < globalBlockHash->hashSize; counter++) {
    uint32_t offset;
    offset = read32bitNumber(globalBlockHash->block, 16 + (4 * counter),
                             globalBlockHash->max);
    if(offset > 0) {
      dumpStrings(globalBlockHash->block, offset, globalBlockHash->max);
    }
  } 
  return 0; // Success
}
  
int main(int argc, char **argv) {
  char *filename, *seek;
  uint8_t *dnsName;
  int32_t dnsLen;
  if(argc > 1) { 
    filename = argv[1];
  } else {
    filename = "bigBlock.bin";
  }
  if(filename[0] == '-' && filename[1] == '-' && filename[2] == 'd') { //--dump
    if(argc > 2) {
      filename = argv[2];
    } else { 
      filename = "bigBlock.bin";
    }
    return dumpContents(filename);
  }  
  if(*filename == '-') {
    printf("blockHashRead v1.0.05\n");
    printf("Usage: blockHashRead {filename} {name to look for}\n");
    printf("Where {name to look for} is a name like 'www.fejs.ml'.\n");
    printf("\n");
    printf("Also: blockHackRead --dump {filename}\n");
    printf("Using --dump will list all of the names in a block hash file\n");
    return 0;
  }
  if(argc > 2) {
    seek = argv[2];
  } else {
    seek = "www.fejs.ml";
  } 
  dnsName = ASCII2DNS(seek,&dnsLen);
  if(dnsName == NULL) { printf("Problem with DNS name\n"); return 1; }
  globalBlockHash = makeBlockHash(filename);
  if(globalBlockHash == NULL) {
    printf("Error reading %s\n",filename);
    return 1;
  }
  printf("%p %d %08x %08x %d\n",globalBlockHash->block, globalBlockHash->max,
      globalBlockHash->sipKey1, globalBlockHash->sipKey2, 
      globalBlockHash->hashSize);
  printf("%s seek result: %d\n",seek,
         blockHasString(globalBlockHash,dnsName,dnsLen));
  return 0;
}
